client/v3/lease.go

```diff
@@ -295,23 +295,39 @@ func (l *lessor) KeepAlive(ctx context.Context, id LeaseID) (<-chan *LeaseKeepAl
 	l.mu.Unlock()
 
 	go l.keepAliveCtxCloser(ctx, id, ka.donec)
+	errorC := make(chan error, 1)
 	l.firstKeepAliveOnce.Do(func() {
-		go l.recvKeepAliveLoop()
+		go l.recvKeepAliveLoop(errorC)
 		go l.deadlineLoop()
 	})
 
+	select {
+	case err := <-errorC:
+		if err != nil {
+			return ch, err
+		}
+	case _, ok := <-ch:
+		if !ok { // The channel has already been closed by l.deadlineLoop()
+			return ch, rpctypes.ErrTimeoutDueToConnectionLost
+		}
+	}
+
 	return ch, nil
 }
 
 func (l *lessor) KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAliveResponse, error) {
 	for {
-		resp, err := l.keepAliveOnce(ctx, id)
+		newCtx, cancel := context.WithTimeout(ctx, l.firstKeepAliveTimeout)
```

Comment:
defer cancel()

Reply:
@serathius We can't add` defer cancel() ` in a for loop, otherwise it may lead to goroutine leak.  I already added cancel() at line 327 and 330.

Reply:
Actually I'm not sure if the timeout should be declared here or outside the retry loop. 
Wouldn't we prefer to stop retrying after the timeout even if the request consecutively fails for the 'transient issues' ?

Reply:
@ptabor  It seems that we have no choice but to declare the timeout inside the for loop, otherwise the l.keepAliveOnce() will be blocked forever when the etcdserver isn't reachable. 

With regard to the retry logic, I just keep it as it's.  The current implementation only tries on certain scenarios.  
When the context timeout, it will stop retrying, please see [client.go#L526](https://github.com/etcd-io/etcd/blob/e433d12656c5dbd41f4f6b085ced134647ffeb14/client/v3/client.go#L526)

Reply:
@ahrtr Why the method would be blocked forever ? I might be missing something.

I though about:
```
newCtx, cancel := context.WithTimeout(ctx, l.firstKeepAliveTimeout)
defer cancel()
while newCtx.Err == nil {  // it might be redundant to the isHaltErr check... but IMHO more readable
    resp, err := l.keepAliveOnce(newCtx, id)
    ...
}
```

Reply:
@ptabor Thanks.  

Yes, the change you proposed is more readable; but the logic might not be correct. Each time when trying, we need to set the new deadline, but your change always keeps using the same deadline/timeout value. So I create the newCtx inside the for loop.

Secondly, yes, you are correct that the check "newCtx.Err() == nil" is redundant to the isHaltErr. Please see [client.go#L526](https://github.com/etcd-io/etcd/blob/e433d12656c5dbd41f4f6b085ced134647ffeb14/client/v3/client.go#L526).

Thirdly, with regard to "Why the method would be blocked forever", because The reason is that the client side keeps waiting for the gRPC to be ready, Please see [options.go#L29](https://github.com/etcd-io/etcd/blob/69279532f46cb7491bb60a588be534100fdc058d/client/v3/options.go#L29). Of course, the change you proposed resolves it as well. 

PS. I just made a minor change, the isHaltErr should check the newCtx instead of the old ctx. Also rebased this PR. 

What do you think?

Reply:
@ptabor  @serathius I did a little refactoring of the method. I added an anonymous function, so that I can call `defer cancel() `directly. PTAL, thx.

Reply:
There are some test failures, please hold on reviewing this PR. Thanks.

Reply:
It took a couple of days to fix all the test failures.